sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -470,3 +397,449 @@ object JoinReorderDPFilters extends PredicateHelper {
+
+    val chromos: Seq[Chromosome] = Seq.fill(determinePopSize(conf, itemsMap.size)) {
+      Chromosome(conf, shuffle(itemsMap), conditions, topOutputSet)
+    }
+
+    new Population(conf, chromos)
+  }
+
+  private def determinePopSize(conf: SQLConf, numRelations: Int): Int = {
+    val relaxFactor = conf.joinReorderGARelaxFactor
+    // The default population size:
+    // # of relations | pop size (RF=3) | pop size (RF=3.5)| pop size  (RF=4)
+    //  < 13          |   DP based      |   DP based       |   DP based
+    //    13          |   20            |   16 (13<16)     |   16
+    //    14          |   25            |   16             |   16
+    //    15          |   32            |   19             |   16
+    //    16          |   40            |   23             |   16
+    //    17          |   50            |   28             |   19
+    //    18          |   64            |   35             |   22
+    //    19          |   80            |   43             |   26
+    //    20          |   101           |   52             |   32
+    //    21          |   128           |   64             |   38
+    //    22          |   128           |   78             |   45
+    //    23          |   128           |   95             |   53
+    //    24          |   128           |   115            |   64
+    //    25          |   128           |   128            |   76
+    //    26          |   128           |   128            |   90
+    //    27          |   128           |   128            |   90
+    //    28          |   128           |   128            |   128
+    //  > 28          |   128           |   128            |   128
+    val size = math.pow(2.0, numRelations / relaxFactor)
+    val max = conf.joinReorderGAMaxPoPSize
+    val min = conf.joinReorderGAMinPoPSize
+
+    math.ceil(math.max(math.min(max, size), min)).toInt
+  }
+
+  private def shuffle(itemsMap: Map[Int, JoinPlan]) : Seq[JoinPlan] = {
+    util.Random.shuffle(itemsMap.values).toSeq
+  }
+}
+
+class Population(conf: SQLConf, val chromos: Seq[Chromosome]) extends Logging {
+  private def sort(): Seq[Chromosome] = {
+    chromos.sortWith((left, right) => left.fitness > right.fitness)
+  }
+
+  def evolve: Population = {
+    // Sort chromos in the population first.
+    var tempChromos = sort()
+    // Begin iteration.
+    val generations = chromos.size
+    for (i <- 1 to generations) {
+      val father = JoinReorderUtils.select(conf, tempChromos, None)
+      val mother = JoinReorderUtils.select(conf, tempChromos, Some(father))
+      val kid = EdgeRecombination.newChromo(father, mother)
+      tempChromos = putToPop(kid, tempChromos)
+      logDebug(s"Iteration $i, fitness for kid: ${kid.fitness}," +
+          s" and Fitness for plans: ${ tempChromos.map(c => c.fitness)}")
+    }
+    new Population(conf, tempChromos)
+  }
+
+  private def putToPop(kid: Chromosome, chromos: Seq[Chromosome]): Seq[Chromosome] = {
+    val tmp = mutable.Buffer[Chromosome]()
```

Comment:
This can be done more elegantly if done in this way:
1. find the index to insert kid, possibly using `indexWhere` given that `chromos` is already sorted;
2. insert the kid into the seq
3. remove the last.

Reply:
Updated, pls have a look.